File: daily/Java/src/Perceptron.java
import java.util.List;
import java.util.Random;
import java.util.Vector;
import java.util.stream.IntStream;


public class Perceptron {
    private static Random rnd = new Random();
    public Vector<Double> Weight;
    public Double Bias;
    private int Size;

    public Perceptron(int size){
        Size = size;
        Weight = new Vector<>(Size);
        IntStream.range(0, size).forEach(
                i-> {
                    Weight.add(rnd.nextDouble());
                }
        );
        Bias = rnd.nextDouble();
    }

    public Double Forward(Vector<Double> input) {
        return Math.signum(IntStream.range(0, Size).mapToDouble(i->Weight.get(i)*input.get(i)).sum()+Bias);
    }
    public void Renew(Vector<Double> input, double target){
        double learningRate = 0.3;
        Renew(input, target, learningRate);
    }
    public void Renew(Vector<Double> input, double target, double learningRate){
        if (Forward(input)*target<0){
            IntStream.range(0, Size).forEach(i->{
                Weight.set(i, Weight.get(i) + learningRate * target * input.get(i));
            });
            Bias = learningRate*target + Bias;
        }
    }

}

File: daily/Kotlin/src/Perceptron.kt
import java.util.*
import java.util.stream.IntStream
import kotlin.streams.toList
typealias Vector = List<Double>
typealias DataSets = List<Pair<Vector, Double>>

object Perceptron {
    val rnd = Random()
    fun dataGen(size: Int, featureLength: Int, f: (Vector) -> Double) =
            IntStream.range(0, size).toList().map {
                val features =
                        IntStream
                                .range(0, featureLength)
                                .toList()
                                .map { rnd.nextDouble() }
                                .toList()

                val d = Math.signum(f(features))
                Pair(features, d)
            }


    class Perceptron(size: Int) {
        var Weight: Vector = rnd.doubles(size.toLong(), 0.0, 1.0).toList()
        var Bias: Double = rnd.nextDouble()
        val Size = size
        fun Forward(input: Vector): Double = Math.signum(Weight.zip(input).sumByDouble { it.first + it.second } + Bias)
        fun Renew(input: Vector, target: Double, learningRate: Double = 0.3) {
            if (Forward(input) * target < 0) {
                IntStream.range(0, Size).forEach {
                    Weight = Weight.zip(input).map { it.first + learningRate * target * it.second }
                    Bias += learningRate * target
                }
            }
        }
    }


}
fun main(args: Array<String>) {
    val datas: DataSets = Perceptron.dataGen(100, 2, { it[0] * 2 + it[1] - 1.5 })
    val perceptron = Perceptron.Perceptron(2)
    IntStream.range(0, 3).forEach {
        for ((data, target) in datas) {
            perceptron.Renew(data, target)
        }
    }
    datas.forEach {
        val (data, target) = it
        println(message = """${perceptron.Forward(data)} $target""")
    }
}
